File: src/lib/api/collection.ts
// Mirrors src-tauri/src/commands/collection.rs -- see lib/api/_PURPOSE.md:
// this is the ONLY place the frontend is allowed to call invoke() for
// collection-tracking data. Tauri camelCases command argument keys
// automatically, so calls here pass camelCase even though the Rust params
// and the returned row shapes (manual_seed.ts) are snake_case.

import { invoke } from "@tauri-apps/api/core";
import type {
  UnitOwnershipOverride,
  UserCollectionEntry,
  UserUnitOwnership,
} from "../types/manual_seed";

export function addOrUpdateOwnedExpansion(
  userId: string,
  expansionId: string,
  quantityOwned: number,
  notes?: string
): Promise<UserCollectionEntry> {
  return invoke<UserCollectionEntry>("add_or_update_owned_expansion", {
    userId,
    expansionId,
    quantityOwned,
    notes: notes ?? null,
  });
}

export function removeOwnedExpansion(userId: string, expansionId: string): Promise<void> {
  return invoke<void>("remove_owned_expansion", { userId, expansionId });
}

export function listOwnedExpansions(userId: string): Promise<UserCollectionEntry[]> {
  return invoke<UserCollectionEntry[]>("list_owned_expansions", { userId });
}

export function setUnitOverride(
  userId: string,
  unitId: string,
  delta: number,
  reason?: string
): Promise<UnitOwnershipOverride> {
  return invoke<UnitOwnershipOverride>("set_unit_override", {
    userId,
    unitId,
    delta,
    reason: reason ?? null,
  });
}

export function removeUnitOverride(userId: string, unitId: string): Promise<void> {
  return invoke<void>("remove_unit_override", { userId, unitId });
}

export function listUnitOverrides(userId: string): Promise<UnitOwnershipOverride[]> {
  return invoke<UnitOwnershipOverride[]>("list_unit_overrides", { userId });
}

export function getUserUnitOwnership(userId: string): Promise<UserUnitOwnership[]> {
  return invoke<UserUnitOwnership[]>("get_user_unit_ownership", { userId });
}

File: src-tauri/src/commands/collection.rs
//! Thin Tauri wrappers over db::queries::collection -- see
//! commands/_PURPOSE.md. No SQL and no business logic here; collection
//! tracking has none beyond what the schema's own CHECK constraints
//! enforce (see the design note in migrations/0002_collection.sql), so
//! there's no domain/ call to make either.

use crate::db::{queries, DbState};
use crate::error::AppError;
use crate::types::{UnitOwnershipOverride, UserCollectionEntry, UserUnitOwnership};
use tauri::State;

#[tauri::command]
pub fn add_or_update_owned_expansion(
    state: State<DbState>,
    user_id: String,
    expansion_id: String,
    quantity_owned: u32,
    notes: Option<String>,
) -> Result<UserCollectionEntry, AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::add_or_update_owned_expansion(
        &conn,
        &user_id,
        &expansion_id,
        quantity_owned,
        notes.as_deref(),
    )?)
}

#[tauri::command]
pub fn remove_owned_expansion(
    state: State<DbState>,
    user_id: String,
    expansion_id: String,
) -> Result<(), AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::remove_owned_expansion(
        &conn,
        &user_id,
        &expansion_id,
    )?)
}

#[tauri::command]
pub fn list_owned_expansions(
    state: State<DbState>,
    user_id: String,
) -> Result<Vec<UserCollectionEntry>, AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::list_owned_expansions(&conn, &user_id)?)
}

#[tauri::command]
pub fn set_unit_override(
    state: State<DbState>,
    user_id: String,
    unit_id: String,
    delta: i32,
    reason: Option<String>,
) -> Result<UnitOwnershipOverride, AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::set_unit_override(
        &conn,
        &user_id,
        &unit_id,
        delta,
        reason.as_deref(),
    )?)
}

#[tauri::command]
pub fn remove_unit_override(
    state: State<DbState>,
    user_id: String,
    unit_id: String,
) -> Result<(), AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::remove_unit_override(
        &conn, &user_id, &unit_id,
    )?)
}

#[tauri::command]
pub fn list_unit_overrides(
    state: State<DbState>,
    user_id: String,
) -> Result<Vec<UnitOwnershipOverride>, AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::list_unit_overrides(&conn, &user_id)?)
}

#[tauri::command]
pub fn get_user_unit_ownership(
    state: State<DbState>,
    user_id: String,
) -> Result<Vec<UserUnitOwnership>, AppError> {
    let conn = state.0.lock().expect("db mutex poisoned");
    Ok(queries::collection::get_user_unit_ownership(
        &conn, &user_id,
    )?)
}
